### DISCUSSION TITLE
Type of Prisma Client with extensions

### DISCUSSION THREAD

**Discussion by joulev** [prisma/prisma#18014] [Q&A] [answered]
I use Next.js so has been initialising the Prisma Client like this

```ts
const globalForPrisma = global as unknown as { prisma: PrismaClient };
export const prisma = globalForPrisma.prisma || new PrismaClient();
if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
```

which works well. But recently I got access to Accelerate, so I changed it to this

```ts
const globalForPrisma = global as unknown as { prisma: PrismaClient };
export const prisma = globalForPrisma.prisma || new PrismaClient().$extends(useAccelerate);
if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
```

But since `prisma` is still `PrismaClient` (not extended), new properties like `cacheStrategy` is not available in the TypeScript type. They still work as expected, functionally, but I cannot get TypeScript to be happy. While I *can* use the following way

```ts
const currentPrisma = new PrismaClient().$extends(useAccelerate);
const globalForPrisma = global as unknown as { prisma: typeof currentPrisma };
export const prisma = globalForPrisma.prisma || currentPrisma;
if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
```

if I use that, certainly things like #4399 will show up.

I also tried to see the type of `currentPrisma` and use it directly instead of `typeof`, but the type is...

```ts
GetClient<PrismaClient<Prisma.PrismaClientOptions, never, Prisma.RejectOnNotFound | Prisma.RejectPerOperation | undefined, {
    result: {} & Record<string, {}>;
    model: {} & ... 1 more ... & {
        ...;
    };
    client: {};
    query: {};
}>, {}>
```

my editor won't even agree to displaying it in full, also even if I could get the type in full, the code would be less than readable.

How can I get both Next.js and TypeScript to work happily while using extensions such as Accelerate?

**Comment by SevInf** [comment, marked as answer]
Hi @joulev. There is definitely a room for improvement from our side here, but this should work for now:

```ts
function createPrisma() {
  return new PrismaClient().$extends(useAccelerate);
}
const globalForPrisma = global as unknown as { prisma: ReturnType<typeof createPrisma> };
export const prisma = globalForPrisma.prisma || createPrisma();
if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
